Support more voices (#3)

oli-ver opened the pull request:
create_voices() function creates a voices list based on the parameters
The list contains the voice name, the tracks for this voice and the instrument
In add_solo_track_bundles now this voices list is used and not sung_parts

oli-ver committed:
Support more voices (#3)

create_voices() function creates a voices list based on the parameters
The list contains the voice name, the tracks for this voice and the instrument
In add_solo_track_bundles now this voices list is used and not sung_parts

oli-ver commented:
@jackkinsella I have a working example of how we could support more voices. The solution allows to specify all kinds of voices and combinations, i. e.

SATB --> # should work with default settings
SSAATTBB --> --common-solo-tracks 0 --soprano 1 2 --alt 3 4 --tenor 5 6 --bass 7 8
SSAA -->  --common-solo-tracks 0 --soprano 1 2 --alt 3 4 --tenor --bass
TTBB --> --common-solo-tracks 0 --soprano --alt --tenor 1 2 --bass 3 4

Currently every voice gets a number, even if there is only 1 (so "soprano 1" for SATB). If that bothers us I could still edit the voices list, but I wanted to check with you first that this is the right approach.

jackkinsella commented:
This is gonna be somewhat dependent on the fix for #1 since I've introduced logic for the tempo map track. The good news is that, soon, I hope the code will be able to auto-detect whether it needs what used by called the "common-solo-track", which will clean up the API a lot.

jackkinsella commented:
As for the API you suggested:

SATB --> # should work with default settings

=> +1
SSAATTBB --> --common-solo-tracks 0 --soprano 1 2 --alt 3 4 --tenor 5 6 --bass 7 8

=> +1, that's exactly how I imagined it should be

SSAA --> --common-solo-tracks 0 --soprano 1 2 --alt 3 4 --tenor --bass
TTBB --> --common-solo-tracks 0 --soprano --alt --tenor 1 2 --bass 3 4

=> In these two cases, would it be possible to leave out the un-used tracks? E.g. SSAA would become `--soprano 1 2 --alt 3 4` instead of `--soprano 1 2 --alt 3 4 --tenor --bass`

oli-ver commented:
> => In these two cases, would it be possible to leave out the un-used tracks? E.g. SSAA would become `--soprano 1 2 --alt 3 4` instead of `--soprano 1 2 --alt 3 4 --tenor --bass`

Only if we remove the defaults for SATB in the argument parser. Right now we need to overwrite the SATB default settings for `--tenor` and `--bass`.

jackkinsella commented:
I see.

 How about this: Could we remove the defaults from the arguments parser, like you say, but then have some second order functionality that intelligently figures out what tracks map to what voices based on the minimal input given to the CLI?

oli-ver committed:
Review Task - Set defaults

The defaults are now set to tracks 1-4 for the voices only if no voice argument was provided

oli-ver commented:
> I see.
> 
> How about this: Could we remove the defaults from the arguments parser, like you say, but then have some second order functionality that intelligently figures out what tracks map to what voices based on the minimal input given to the CLI?

Good idea, I just changed that. Is this how you meant that?

jackkinsella reviewed (changes requested):
The functionality all looks good. The only thing I'll request before merging this PR is that you add at least one regression test of the new CLI usage e.g. `--soprano 1 2 --alto 3 4`. See the example test I added in `tests`.

The other thing, which we can discuss and do in a future separate PR, is cleaning up the names and refactoring into more structured types (e.g. Voice vs Part).

oli-ver commented on midi-to-part-mp3s:
```diff
@@ -180,19 +190,51 @@ def separate_tracks_into_mp3s(args: argparse.Namespace, midifile_path: str) -> N
+  Returns:
+      List[str][int] -- list of voices and their assigned midi track numbers
+  """
+  voices: List[List[str][int]] = []
```
Ok, nested lists are quite common in Python. But I am not sure if the type hint is correct this way. A list can always contain n elements. I will check.

oli-ver replied:
Ok, nested lists are quite common in Python. But I am not sure if the type hint is correct this way. A list can always contain n elements. I will check.

oli-ver replied:
I changed that to List[List] to show that's a nested list. I don't think the previous way was even possible, I did not find it anywhere in the docs.

oli-ver commented on midi-to-part-mp3s:
```diff
@@ -180,19 +190,51 @@ def separate_tracks_into_mp3s(args: argparse.Namespace, midifile_path: str) -> N
+  """
+  voices: List[List[str][int]] = []
+  sung_part: str
+  for sung_part in sung_parts:
```
voice_midi_track could be better: it's the configured track number for the voice

oli-ver replied:
voice_midi_track could be better: it's the configured track number for the voice

oli-ver replied:
I used voice_midi_tracks because it's a list actually

oli-ver commented on midi-to-part-mp3s:
```diff
@@ -124,12 +124,21 @@ def convert_midi_to_mp3(midifile_path: str) -> None:
+  """
+  track: mido.MidiTrack
   for track in midi_data.tracks:
+    message: mido.Message
```
No but in VSCode it is used for code completion. I think if I don't input it typing "message." will not list methods and fields.

oli-ver commented on midi-to-part-mp3s:
```diff
@@ -86,25 +86,25 @@ def instrument_number_for_part(part: str) -> int:
+  midi: mido.MidiFile = mido.MidiFile()
+  midi.ticks_per_beat: int = midi_data.ticks_per_beat
 
+  track_number:int
```
No, it's true. I am currently not using an autoformatter. Are you? Formatting the code with autopep8 changes pretty much everything. I will change that.

oli-ver commented on midi-to-part-mp3s:
```diff
@@ -86,25 +86,25 @@ def instrument_number_for_part(part: str) -> int:
 
-  if part_name in sung_parts:
-    instrument_number = instrument_number_for_part(part_name)
+  if part_name != 'accompaniment':
```
Yes, I think we could optimize that :)

oli-ver commented on midi-to-part-mp3s:
```diff
@@ -180,19 +190,51 @@ def separate_tracks_into_mp3s(args: argparse.Namespace, midifile_path: str) -> N
+        track_id = voice_parameter[i]
+        tracks.append(track_id)
+        instrument = instrument_number_for_part(sung_part)
+        voice = [sung_part + ' ' + str(i+1), tracks, instrument]
```
mmh you are right, we could perhaps just add the needed fields to the Parts class

oli-ver commented on midi-to-part-mp3s:
```diff
@@ -180,19 +190,51 @@ def separate_tracks_into_mp3s(args: argparse.Namespace, midifile_path: str) -> N
+
+
+
+def add_solo_track_bundles(voices: list):
```
With the change that there is no `--common-solo-track` anymore this is now just the method to add accompaniment tracks. I will update the description and refactor the name to `add_accompaniment()`

oli-ver commented on midi-to-part-mp3s:
```diff
@@ -203,14 +245,37 @@ def remove_temporary_midifiles() -> None:
+  """Sets the soprano, alto, tenor and bass voices 
+  to the default tracks 1, 2, 3 and 4, if no track has been defined.
+  This allows a SSAA or TTBB API setting without the need to overwrite
+  defaults, i. e.:
```
I changed that

oli-ver committed:
Review Results

Changed some formattings
Refactored  add_solo_track_bundles() to add_accompaniment() to better reflect what the function actually does

oli-ver commented:
> The functionality all looks good. The only thing I'll request before merging this PR is that you add at least one regression test of the new CLI usage e.g. `--soprano 1 2 --alto 3 4`. See the example test I added in `tests`.
> 
> The other thing, which we can discuss and do in a future separate PR, is cleaning up the names and refactoring into more structured types (e.g. Voice vs Part).

I worked on some of the review topics and wanted to create a test for the new feature. Before I was able to begin I noticed that the regression test you added fails on my branch:

```
AssertionError: Sequences differ: [11.310998, 10.867007, 10.867007, 10.867007] != [11.310998, 11.310998, 11.310998, 11.310998]

First differing element 1:
10.867007
11.310998

- [11.310998, 10.867007, 10.867007, 10.867007]
+ [11.310998, 11.310998, 11.310998, 11.310998]

----------------------------------------------------------------------
Ran 2 tests in 31.780s

FAILED (failures=1)
```

That's strange to me as I did not change anything about the way you are editing the tracks to fix the timing issues. I will have a deeper look in the next days.

oli-ver committed:
Fixed Unit test file names

The file names needed to be fixed. The unit test fails right now.

oli-ver committed:
Fix tempo map not included

Refactored the function solo_track_bundles() into create_voices() to include the tempo map track again

oli-ver commented:
Found it! This shows your regression test works very well :) I did not include the tempo map in the voices. I will work on the regression test now to make sure the files are created correctly, so i.e. if `--soprano 1 2` has been configured no other voices should be created, the default settings create default SATB and so on.

oli-ver committed:
Added unittest test_more_voices

Changed filename of main file: minus sign not allowed in python file names and therefore only importable using workarounds
Made args global to fix the todo that args is being passed too deep into the call stack
Created main method to be able to call the script from outside after import
Created unittest for SATB, SSAA and SSAATTBB
Added Abendlied-Rheinberger.mxl to implicitely test mxl midi conversion and SSAATTBB

oli-ver committed:
Allow calling the app with the command midi-to-part-mp3s

The short script midi-to-part-mp3s makes it possible again to call the application the old way without having the import issues with the main module

oli-ver commented:
During working on the unittest I stumbled across some issues with the file name. I wanted to import the module and call it's main method rather than calling it from the shell like you did. Therefore I had to change the filename on the way and change the application initilization a little bit. I also added a convenience script with the old name.

The changes will allow us to call separate methods of the script in future unit tests. Right now we both only developed end-to-end / integration tests, but this will probably change at some point i.e. if we only want to unittest a single method

jackkinsella commented on midi_to_part_mp3s.py:
```diff
@@ -20,6 +21,7 @@
   "tenor": 49, # string ensemble 1
   "bass": 50 # string ensemble 2
 }
+args = None
```
Thanks, big improvement in cleaner method sigs!

jackkinsella reviewed (approved):
Excellent stuff and just in time — I need to convert a SATTB track today.

Abendlied is a lovely choral piece. Our test fixtures have been thoughtfully curated :)
